File: dsa-problems/Probabilistic/solutions/javascript/PRB-013-random-walk-hitting-prob-2d.js
const readline = require("readline");

function hitProbability(a, b, T) {
  if (a === 0 && b === 0) return 1.0;
  
  const offset = T;
  const size = 2 * T + 1;
  let dp = new Float64Array(size * size);
  
  const targetX = a + offset;
  const targetY = b + offset;
  
  // Helper for 2D index
  const idx = (x, y) => x * size + y;
  
  dp[idx(offset, offset)] = 1.0;
  
  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];
  
  for (let t = 1; t <= T; t++) {
    const nextDp = new Float64Array(size * size);
    
    // Carry over absorbed probability
    nextDp[idx(targetX, targetY)] = dp[idx(targetX, targetY)];
    
    const minVal = Math.max(0, offset - (t - 1));
    const maxVal = Math.min(size - 1, offset + (t - 1));
    
    for (let x = minVal; x <= maxVal; x++) {
      for (let y = minVal; y <= maxVal; y++) {
        const currIdx = idx(x, y);
        const val = dp[currIdx];
        
        if (val === 0) continue;
        if (x === targetX && y === targetY) continue;
        
        const prob = val * 0.25;
        for (let i = 0; i < 4; i++) {
          const nx = x + dx[i];
          const ny = y + dy[i];
          const nIdx = idx(nx, ny);
          nextDp[nIdx] += prob;
        }
      }
    }
    dp = nextDp;
  }
  
  return dp[idx(targetX, targetY)];
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  const a = parseInt(data[0], 10);
  const b = parseInt(data[1], 10);
  const T = parseInt(data[2], 10);
  console.log(hitProbability(a, b, T).toFixed(6));
});

File: dsa-problems/Probabilistic/solutions/cpp/PRB-013-random-walk-hitting-prob-2d.cpp
#include <iostream>
#include <vector>
#include <iomanip>

using namespace std;

class Solution {
public:
    double hitProbability(int a, int b, int T) {
        if (a == 0 && b == 0) return 1.0;
        
        int offset = T;
        int size = 2 * T + 1;
        
        // Use vector of vectors
        vector<vector<double>> dp(size, vector<double>(size, 0.0));
        
        int targetX = a + offset;
        int targetY = b + offset;
        
        dp[offset][offset] = 1.0;
        
        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};
        
        for (int t = 1; t <= T; t++) {
            vector<vector<double>> nextDp(size, vector<double>(size, 0.0));
            
            nextDp[targetX][targetY] = dp[targetX][targetY];
            
            int minVal = max(0, offset - (t - 1));
            int maxVal = min(size - 1, offset + (t - 1));
            
            for (int x = minVal; x <= maxVal; x++) {
                for (int y = minVal; y <= maxVal; y++) {
                    if (dp[x][y] == 0) continue;
                    if (x == targetX && y == targetY) continue;
                    
                    double prob = dp[x][y] * 0.25;
                    for (int i = 0; i < 4; i++) {
                        int nx = x + dx[i];
                        int ny = y + dy[i];
                        nextDp[nx][ny] += prob;
                    }
                }
            }
            dp = nextDp;
        }
        
        return dp[targetX][targetY];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int a, b, T;
    if (cin >> a >> b >> T) {
        Solution solution;
        cout << fixed << setprecision(6) << solution.hitProbability(a, b, T) << "\n";
    }
    return 0;
}
